File: Ahorn/entities/CoreFragment.jl
module CoreFragment

using ..Ahorn, Maple

@mapdef Entity "MaggyHelper/CoreFragment" CoreFragment(x::Integer, y::Integer, fragmentId::String="", fragmentIndex::Integer=1, requiredShields::Integer=3, requiresProtection::Bool=true)

const placements = Ahorn.PlacementDict(
    "protected" => Ahorn.EntityPlacement(CoreFragment),
    "open" => Ahorn.EntityPlacement(CoreFragment)
)

function Ahorn.selection(entity::CoreFragment)
    return Ahorn.Rectangle(entity.x - 8, entity.y - 8, 16, 16)
end

function Ahorn.render(ctx::Ahorn.Cairo.CairoContext, entity::CoreFragment, room::Maple.Room)
    Ahorn.drawRectangle(ctx, Ahorn.Rectangle(entity.x - 8, entity.y - 8, 16, 16), (0.5, 0.5, 0.5, 0.5), (0.0, 0.0, 0.0, 1.0))
end

end

File: framework/extracted/Loenn/entities/coreFragment.lua
local utils = require("utils")
local drawableSprite = require("structs.drawable_sprite")

local coreFragment = {}

coreFragment.name = "MaggyHelper/CoreFragment"
coreFragment.depth = -100
coreFragment.fieldInformation = {
    fragmentIndex = {
        minimumValue = 1
    },
    requiredShields = {
        minimumValue = 0
    }
}
coreFragment.placements = {
    {
        name = "protected",
        data = {
            fragmentId = "",
            fragmentIndex = 1,
            requiresProtection = true,
            requiredShields = 3
        }
    },
    {
        name = "open",
        data = {
            fragmentId = "",
            fragmentIndex = 1,
            requiresProtection = false,
            requiredShields = 0
        }
    }
}

function coreFragment.sprite(room, entity)
    local state = entity.requiresProtection and "protected" or "active"
    local sprite = drawableSprite.fromTexture("objects/core_fragment/" .. state, entity)
    sprite:setPosition(entity.x, entity.y)
    sprite:setJustification(0.5, 0.5)
    
    -- Color based on index
    local colors = {
        {0.0, 1.0, 1.0, 1.0},
        {0.0, 1.0, 0.5, 1.0},
        {1.0, 0.5, 0.0, 1.0},
        {1.0, 0.0, 0.5, 1.0}
    }
    local idx = (entity.fragmentIndex or 1) % 4
    if idx == 0 then idx = 4 end
    sprite.color = colors[idx]
    
    return sprite
end

function coreFragment.selection(room, entity)
    return utils.rectangle(entity.x - 12, entity.y - 12, 24, 24)
end

return coreFragment
